src/main/java/org/terasology/dynamicCities/buildings/BuildingManager.java

```diff
@@ -240,37 +228,42 @@ public void postBegin() {
         }
     }
 
-    public Optional<List<EntityRef>> getTemplatesForBuilding(GenericBuildingComponent building) {
-        List<EntityRef> templateList = new ArrayList<>();
-
-        if (building.templateNames == null) {
-            return Optional.empty();
-        }
-
-        for (String name : building.templateNames) {
-            Optional<EntityRef> templateOptional = getTemplate(name);
-            if (templateOptional.isPresent()) {
-                templateList.add(templateOptional.get());
-            }
+    /**
+     * Attempt to resolve the given {@code names} with the given {@code resolver}.
+     *
+     * @param names the names to resolve with the resolver; may be null
+     * @param resolver function to resolve an object of type {@code T} from a name
+     * @param <T> the type of the object to resolve
+     * @return {@link Optional#empty()} if {@code names} is null, or some list containing the resolved objects.
+     */
+    static <T> Optional<List<T>> resolveFromNames(List<String> names, Function<String, Optional<T>> resolver) {
+        return Optional.ofNullable(names)
+                .map(templateNames ->
+                        templateNames.stream()
+                                .map(resolver)
+                                .filter(Optional::isPresent)
+                                .map(Optional::get)
```

Comment:
I wish we were using Java 9 so instead of `.filter().map()` it could be

```java
templateNames.stream()
    .map(resolver)
    .flatMap(Optional::stream)
```

I learned there's actually `Streams::stream` for that purpose in Guava, but that's marked beta and this isn't actually a big deal.

Reply:
Yeah, I know 🙈 Should probably add a comment here marking this as a potential Java 9+ improvement.